diff --git a/python/torch_mlir/extras/fx_decomp_util.py b/python/torch_mlir/extras/fx_decomp_util.py
index 8dddede2..43d79933 100644
--- a/python/torch_mlir/extras/fx_decomp_util.py
+++ b/python/torch_mlir/extras/fx_decomp_util.py
@@ -1,6 +1,11 @@
+from packaging import version
+
 import torch
 from torch._decomp import get_decompositions
 
+torch_version = version.Version(torch.__version__)
+TORCH_VERSION_BIGGER_THAN_2_3_0 = torch_version > version.Version("2.3.0")
+
 # default decompositions pulled from SHARK / torch._decomp
 DEFAULT_DECOMPOSITIONS = [
     torch.ops.aten.embedding_dense_backward,
@@ -48,10 +53,12 @@ DEFAULT_DECOMPOSITIONS = [
     torch.ops.aten.triu.default,
     torch.ops.aten.nan_to_num.default,
     torch.ops.aten.unbind,
-    torch.ops.aten._scaled_dot_product_flash_attention_for_cpu,
     torch.ops.aten.diag,
 ]
 
+if TORCH_VERSION_BIGGER_THAN_2_3_0:
+    DEFAULT_DECOMPOSITIONS += [torch.ops.aten._scaled_dot_product_flash_attention_for_cpu]
+
 
 def get_decomposition_table():
     return get_decompositions(DEFAULT_DECOMPOSITIONS)
