diff --git a/lib/Dialect/Torch/IR/TorchOps.cpp b/lib/Dialect/Torch/IR/TorchOps.cpp
index 5837e745..eab6c5b2 100644
--- a/lib/Dialect/Torch/IR/TorchOps.cpp
+++ b/lib/Dialect/Torch/IR/TorchOps.cpp
@@ -3088,6 +3088,9 @@ void PrimListUnpackOp::getCanonicalizationPatterns(RewritePatternSet &patterns,
     if (!listConstruct)
       return failure();
 
+    if (op->getNumResults() != listConstruct.getElements().size())
+      return failure();
+
     rewriter.replaceOp(op, listConstruct.getElements());
     return success();
   });
