diff --git a/lib/Dialect/Torch/Transforms/ReduceOpVariants.cpp b/lib/Dialect/Torch/Transforms/ReduceOpVariants.cpp
index 84780e04..b2617737 100644
--- a/lib/Dialect/Torch/Transforms/ReduceOpVariants.cpp
+++ b/lib/Dialect/Torch/Transforms/ReduceOpVariants.cpp
@@ -444,9 +444,7 @@ struct ReduceOpVariantsPass
         }
       }
       if (op->hasTrait<Torch::OpTrait::HasValueSemantics>() ||
-          (isa<OperatorOp>(op) &&
-           operatorOpHasValueSemantics(cast<OperatorOp>(op),
-                                       extraLibraryModuleSymTable))) {
+          isa<OperatorOp>(op)) {
         auto hasValueSemantics = [](Type t) {
           // TODO: Make this an allowlist based on a closed torch dialect
           // type system.
