diff --git a/python/aitemplate/testing/detect_target.py b/python/aitemplate/testing/detect_target.py
index 2b2913d..95fd02f 100644
--- a/python/aitemplate/testing/detect_target.py
+++ b/python/aitemplate/testing/detect_target.py
@@ -42,7 +42,7 @@ def _detect_cuda_with_nvidia_smi():
         sm_names = {
             "70": ["V100"],
             "75": ["T4", "Quadro T2000"],
-            "80": ["PG509", "A100", "A10G", "RTX 30", "A30", "RTX 40"],
+            "80": ["PG509", "A100", "A10G", "RTX 30", "A30", "RTX 40", "A10", "A16"],
             "90": ["H100"],
         }
         for sm, names in sm_names.items():
