diff --git a/python/aitemplate/backend/builder.py b/python/aitemplate/backend/builder.py
index e66d97e..b0d5848 100644
--- a/python/aitemplate/backend/builder.py
+++ b/python/aitemplate/backend/builder.py
@@ -900,6 +900,7 @@ clean:
                 f"-C {build_dir}",
             ]
         )
+        self._n_jobs = 4
         make_clean_cmd = f" {make_path} {make_flags} clean "
         make_all_cmd = f" {make_path} {make_flags} -j{self._n_jobs} all "
         make_clean_constants_cmd = f" {make_path} {make_flags} clean_constants "
