diff --git a/tensorflow/compiler/xla/mlir_hlo/mhlo/IR/hlo_ops.cc b/tensorflow/compiler/xla/mlir_hlo/mhlo/IR/hlo_ops.cc
index 8a4bef78a19..92f74d3b7a9 100644
--- a/tensorflow/compiler/xla/mlir_hlo/mhlo/IR/hlo_ops.cc
+++ b/tensorflow/compiler/xla/mlir_hlo/mhlo/IR/hlo_ops.cc
@@ -5650,8 +5650,9 @@ static Attribute CompareFolder(CompareOp op, ArrayRef<Attribute> attrs) {
 
   SmallVector<bool, 6> values;
   values.reserve(lhs.getNumElements());
-  for (const auto zip :
-       llvm::zip(lhs.getValues<SrcType>(), rhs.getValues<SrcType>())) {
+  auto zips = llvm::to_vector(
+      llvm::zip(lhs.getValues<SrcType>(), rhs.getValues<SrcType>()));
+  for (const auto zip : zips) {
     values.push_back(
         Convert()(addSign(std::get<0>(zip), lhs.getElementType()),
                   addSign(std::get<1>(zip), rhs.getElementType())));
