diff --git a/third_party/xla/xla/mlir_hlo/mhlo/IR/hlo_ops.cc b/third_party/xla/xla/mlir_hlo/mhlo/IR/hlo_ops.cc
index ee1e08303e7..4cccf29e6ae 100644
--- a/third_party/xla/xla/mlir_hlo/mhlo/IR/hlo_ops.cc
+++ b/third_party/xla/xla/mlir_hlo/mhlo/IR/hlo_ops.cc
@@ -5758,8 +5758,9 @@ static Attribute CompareFolder(CompareOp op, ArrayRef<Attribute> attrs) {
 
   SmallVector<bool, 6> values;
   values.reserve(lhs.getNumElements());
-  for (const auto zip :
-       llvm::zip(lhs.getValues<SrcType>(), rhs.getValues<SrcType>())) {
+  auto zips = llvm::to_vector(
+      llvm::zip(lhs.getValues<SrcType>(), rhs.getValues<SrcType>()));
+  for (const auto zip : zips) {
     values.push_back(
         Convert()(addSign(std::get<0>(zip), lhs.getElementType()),
                   addSign(std::get<1>(zip), rhs.getElementType())));
