diff --git a/tensorflow/compiler/mlir/tf2xla/transforms/legalize_tf.cc b/tensorflow/compiler/mlir/tf2xla/transforms/legalize_tf.cc
index 13c9c3f9306..fc78a4420a5 100644
--- a/tensorflow/compiler/mlir/tf2xla/transforms/legalize_tf.cc
+++ b/tensorflow/compiler/mlir/tf2xla/transforms/legalize_tf.cc
@@ -2868,6 +2868,9 @@ class ConvertSelectOp : public OpRewritePattern<TF::SelectOp> {
 
   LogicalResult matchAndRewrite(TF::SelectOp op,
                                 PatternRewriter &rewriter) const override {
+    if(op.getOutput().getType().getElementType().isa<mlir::TF::StringType>()) {
+      return failure();
+    }
     // This lowering only works on ranked types.
     auto cond_type =
         mlir::dyn_cast<RankedTensorType>(op.getCondition().getType());
