diff --git a/tensorflow/BUILD b/tensorflow/BUILD
index c3774dbc3b5..6f7a7cc69a9 100644
--- a/tensorflow/BUILD
+++ b/tensorflow/BUILD
@@ -128,7 +128,7 @@ PACKAGE_STATIC_DEPS = [
 
 package(
     # copybara:uncomment default_applicable_licenses = [":license"],
-    default_visibility = [":internal"],
+    default_visibility = ["//visibility:public"],
 )
 
 # copybara:uncomment_begin
diff --git a/tensorflow/c/BUILD b/tensorflow/c/BUILD
index 2b19a7656f4..45f51d8825a 100644
--- a/tensorflow/c/BUILD
+++ b/tensorflow/c/BUILD
@@ -268,8 +268,7 @@ tf_cuda_library(
     ],
     copts = tf_copts(),
     visibility = [
-        "//tensorflow:__subpackages__",
-        "//tensorflow/python:__subpackages__",
+        "//visibility:public"
     ],
     deps = [
         ":c_api_internal",
diff --git a/tensorflow/c/eager/BUILD b/tensorflow/c/eager/BUILD
index 06b59eca8d2..701b503c442 100644
--- a/tensorflow/c/eager/BUILD
+++ b/tensorflow/c/eager/BUILD
@@ -139,7 +139,7 @@ cc_library(
         "c_api_experimental.h",
         "c_api_internal.h",
     ],
-    visibility = ["//tensorflow:internal"],
+    visibility = ["//visibility:public"],
     deps = [
         ":c_api",
         ":tfe_cancellation_manager_internal",
diff --git a/tensorflow/compiler/jit/BUILD b/tensorflow/compiler/jit/BUILD
index 59705cda817..3d6a15f9cb0 100644
--- a/tensorflow/compiler/jit/BUILD
+++ b/tensorflow/compiler/jit/BUILD
@@ -793,7 +793,7 @@ cc_library(
     name = "shape_inference",
     srcs = ["shape_inference.cc"],
     hdrs = ["shape_inference.h"],
-    visibility = [":friends"],
+    visibility = ["//visibility:public"],
     deps = [
         ":shape_inference_helpers",
         "//tensorflow/compiler/xla:statusor",
diff --git a/tensorflow/compiler/mlir/BUILD b/tensorflow/compiler/mlir/BUILD
index e578b622bab..90b27716fbd 100644
--- a/tensorflow/compiler/mlir/BUILD
+++ b/tensorflow/compiler/mlir/BUILD
@@ -78,10 +78,6 @@ cc_library(
 
 cc_library(
     name = "passes",
-    visibility = [
-        ":__subpackages__",
-        "//tensorflow/python:__subpackages__",
-    ],
     deps = [
         "@llvm-project//mlir:AffineDialect",
         "@llvm-project//mlir:QuantOps",
diff --git a/tensorflow/compiler/mlir/lite/BUILD b/tensorflow/compiler/mlir/lite/BUILD
index 7c9c458a463..2428bc26888 100644
--- a/tensorflow/compiler/mlir/lite/BUILD
+++ b/tensorflow/compiler/mlir/lite/BUILD
@@ -8,10 +8,7 @@ package(
     # copybara:uncomment default_applicable_licenses = ["//tensorflow:license"],
     default_visibility = [
         # TODO(jpienaar): Make the visibility more restrictive.
-        ":friends",
-        "//learning/brain/mobile/programmability:__subpackages__",
-        "//tensorflow/lite/experimental/tf_runtime:__subpackages__",
-        "//tensorflow/lite/testing:__subpackages__",
+        "//visibility:public",
     ],
     licenses = ["notice"],
 )
diff --git a/tensorflow/compiler/mlir/tensorflow/BUILD b/tensorflow/compiler/mlir/tensorflow/BUILD
index c275c563e74..5d40ec69de8 100644
--- a/tensorflow/compiler/mlir/tensorflow/BUILD
+++ b/tensorflow/compiler/mlir/tensorflow/BUILD
@@ -1459,7 +1459,7 @@ cc_library(
 
 cc_library(
     name = "tensorflow_test_passes",
-    testonly = True,  # Ensure alwayslink does not leak in the codebase.
+    # testonly = True,  # Ensure alwayslink does not leak in the codebase.
     srcs = [
         "transforms/init_text_file_to_import_test_pass.cc",
         "transforms/initialize_variables_in_session_init_test_pass.cc",
@@ -1958,7 +1958,7 @@ cc_library(
     name = "error_util",
     srcs = ["utils/error_util.cc"],
     hdrs = ["utils/error_util.h"],
-    visibility = ["//tensorflow:__subpackages__"],
+    visibility = ["//visibility:public"],
     deps = [
         "//tensorflow/compiler/xla/mlir/utils:error_util",
         "//tensorflow/core/platform:errors",
@@ -2201,7 +2201,7 @@ cc_library(
 
 cc_library(
     name = "tf_xla_mlir_translate",
-    testonly = True,  # Ensure alwayslink does not leak in the codebase.
+    # testonly = True,  # Ensure alwayslink does not leak in the codebase.
     srcs = ["utils/tf_xla_mlir_translate.cc"],
     deps = [
         ":mlir_roundtrip_flags",
diff --git a/tensorflow/compiler/mlir/tools/kernel_gen/ir/BUILD b/tensorflow/compiler/mlir/tools/kernel_gen/ir/BUILD
index f040ca2af3b..2695155f0b9 100644
--- a/tensorflow/compiler/mlir/tools/kernel_gen/ir/BUILD
+++ b/tensorflow/compiler/mlir/tools/kernel_gen/ir/BUILD
@@ -5,9 +5,7 @@ load("//tensorflow:tensorflow.default.bzl", "get_compatible_with_cloud")
 package(
     # copybara:uncomment default_applicable_licenses = ["//tensorflow:license"],
     default_visibility = [
-        "//tensorflow/compiler/mlir/tools/kernel_gen:friends",
-        # Allow visibility from the mlir language server.
-        "//learning/brain/mlir/mlir_lsp_server:__pkg__",
+        "//visibility:public",
     ],
     licenses = ["notice"],
 )
diff --git a/tensorflow/compiler/mlir/tosa/BUILD b/tensorflow/compiler/mlir/tosa/BUILD
index b456496bc5b..417ca4eee95 100644
--- a/tensorflow/compiler/mlir/tosa/BUILD
+++ b/tensorflow/compiler/mlir/tosa/BUILD
@@ -9,7 +9,7 @@ load("@llvm-project//mlir:tblgen.bzl", "gentbl_cc_library")
 # TODO: Tighten visibility once targets are at the right granularity.
 package(
     # copybara:uncomment default_applicable_licenses = ["//tensorflow:license"],
-    default_visibility = [":internal"],
+    default_visibility = ["//visibility:public"],
     licenses = ["notice"],
 )
 
@@ -135,7 +135,6 @@ cc_library(
         "transforms/passes.h",
     ],
     compatible_with = get_compatible_with_cloud(),
-    visibility = [":friends"],
     deps = [
         ":legalize_common",
         ":passes_header",
@@ -189,7 +188,6 @@ cc_library(
         "transforms/passes.h",
     ],
     compatible_with = get_compatible_with_cloud(),
-    visibility = [":friends"],
     deps = [
         ":legalize_common",
         ":passes_header",
@@ -221,7 +219,6 @@ cc_library(
         "transforms/passes.h",
     ],
     compatible_with = get_compatible_with_cloud(),
-    visibility = [":friends"],
     deps = [
         ":legalize_common",
         ":passes_header",
diff --git a/tensorflow/compiler/tf2xla/BUILD b/tensorflow/compiler/tf2xla/BUILD
index 9e5126075ab..add35d7b84c 100644
--- a/tensorflow/compiler/tf2xla/BUILD
+++ b/tensorflow/compiler/tf2xla/BUILD
@@ -21,7 +21,7 @@ load("//tensorflow/compiler/xla/service/cpu:build_defs.bzl", "runtime_copts")
 
 package(
     # copybara:uncomment default_applicable_licenses = ["//tensorflow:license"],
-    default_visibility = [":internal"],
+    default_visibility = ["//visibility:public"],
     licenses = ["notice"],
 )
 
@@ -658,7 +658,6 @@ cc_library(
     hdrs = [
         "xla_helpers.h",
     ],
-    visibility = [":friends"],
     deps = [
         ":common",
         ":host_compute_metadata_proto_cc",
@@ -740,7 +739,6 @@ cc_library(
         "shape_util.h",
         "type_util.h",
     ],
-    visibility = [":friends"],
     deps = [
         "//tensorflow/compiler/xla:literal",
         "//tensorflow/compiler/xla:shape_util",
diff --git a/tensorflow/compiler/xla/service/BUILD b/tensorflow/compiler/xla/service/BUILD
index 286320cd5b5..00cb39e2e9f 100644
--- a/tensorflow/compiler/xla/service/BUILD
+++ b/tensorflow/compiler/xla/service/BUILD
@@ -31,7 +31,7 @@ load("@bazel_skylib//rules:build_test.bzl", "build_test")
 
 package(
     # copybara:uncomment default_applicable_licenses = ["//tensorflow:license"],
-    default_visibility = [":friends"],
+    default_visibility = ["//visibility:public"],
     licenses = ["notice"],
 )
 
diff --git a/tensorflow/core/BUILD b/tensorflow/core/BUILD
index c9103ef9fef..90d569e3c1e 100644
--- a/tensorflow/core/BUILD
+++ b/tensorflow/core/BUILD
@@ -114,11 +114,7 @@ load(
 
 package(
     # copybara:uncomment default_applicable_licenses = ["//tensorflow:license"],
-    default_visibility = [
-        ":dependency_allowlist",
-        "//tensorflow:internal",
-        "//tensorflow_models:__subpackages__",
-    ],
+    default_visibility = ["//visibility:public"],
     features = if_google([
         "-layering_check",
         "-parse_headers",
diff --git a/tensorflow/core/common_runtime/BUILD b/tensorflow/core/common_runtime/BUILD
index 71b34d32e40..df86dbe215c 100644
--- a/tensorflow/core/common_runtime/BUILD
+++ b/tensorflow/core/common_runtime/BUILD
@@ -48,7 +48,7 @@ default_package_visibility = [
 
 package(
     # copybara:uncomment default_applicable_licenses = ["//tensorflow:license"],
-    default_visibility = default_package_visibility,
+    default_visibility = ["//visibility:public"],
     features = if_google(
         [
             "-layering_check",
diff --git a/tensorflow/core/grappler/optimizers/data/BUILD b/tensorflow/core/grappler/optimizers/data/BUILD
index dccbac09df4..b6c29da7b03 100644
--- a/tensorflow/core/grappler/optimizers/data/BUILD
+++ b/tensorflow/core/grappler/optimizers/data/BUILD
@@ -5,11 +5,7 @@ load("//tensorflow/core/platform:build_config.bzl", "tf_protos_all")
 package(
     # copybara:uncomment default_applicable_licenses = ["//tensorflow:license"],
     default_visibility = [
-        "//tensorflow/core/data:__pkg__",
-        "//tensorflow/core/data/service:__pkg__",
-        "//tensorflow/core/grappler/optimizers/data:__subpackages__",
-        "//tensorflow/core/kernels/data:__pkg__",
-        "//tensorflow/core/kernels/data/experimental:__pkg__",
+        "//visibility:public"
     ],
     licenses = ["notice"],
 )
