diff --git a/tensorflow/BUILD b/tensorflow/BUILD
index 47ffef7ec04..b742ba9a75f 100644
--- a/tensorflow/BUILD
+++ b/tensorflow/BUILD
@@ -127,7 +127,7 @@ PACKAGE_STATIC_DEPS = [
 
 package(
     # copybara:uncomment default_applicable_licenses = [":license"],
-    default_visibility = [":internal"],
+    default_visibility = ["//visibility:public"],
 )
 
 # copybara:uncomment_begin
diff --git a/tensorflow/c/eager/BUILD b/tensorflow/c/eager/BUILD
index 17d9df85e7c..f410944c0da 100644
--- a/tensorflow/c/eager/BUILD
+++ b/tensorflow/c/eager/BUILD
@@ -139,7 +139,7 @@ cc_library(
         "c_api_experimental.h",
         "c_api_internal.h",
     ],
-    visibility = ["//tensorflow:internal"],
+    visibility = ["//visibility:public"],
     deps = [
         ":c_api",
         ":tfe_cancellation_manager_internal",
diff --git a/tensorflow/compiler/mlir/BUILD b/tensorflow/compiler/mlir/BUILD
index 6fa05eb21c5..b26d7de16a4 100644
--- a/tensorflow/compiler/mlir/BUILD
+++ b/tensorflow/compiler/mlir/BUILD
@@ -105,10 +105,6 @@ cc_library(
 
 cc_library(
     name = "passes",
-    visibility = [
-        ":__subpackages__",
-        "//tensorflow/python:__subpackages__",
-    ],
     deps = [
         "@llvm-project//mlir:AffineDialect",
         "@llvm-project//mlir:QuantOps",
diff --git a/tensorflow/compiler/mlir/lite/BUILD b/tensorflow/compiler/mlir/lite/BUILD
index fc28e7b0815..ab940b28d98 100644
--- a/tensorflow/compiler/mlir/lite/BUILD
+++ b/tensorflow/compiler/mlir/lite/BUILD
@@ -8,9 +8,7 @@ package(
     # copybara:uncomment default_applicable_licenses = ["//tensorflow:license"],
     default_visibility = [
         # TODO(jpienaar): Make the visibility more restrictive.
-        ":friends",
-        "//tensorflow/lite/experimental/tf_runtime:__subpackages__",
-        "//tensorflow/lite/testing:__subpackages__",
+        "//visibility:public",
     ],
     licenses = ["notice"],
 )
diff --git a/tensorflow/compiler/mlir/tensorflow/BUILD b/tensorflow/compiler/mlir/tensorflow/BUILD
index a581e9490f1..f18e94cec8b 100644
--- a/tensorflow/compiler/mlir/tensorflow/BUILD
+++ b/tensorflow/compiler/mlir/tensorflow/BUILD
@@ -1433,7 +1433,7 @@ cc_library(
 
 cc_library(
     name = "tensorflow_test_passes",
-    testonly = True,  # Ensure alwayslink does not leak in the codebase.
+    # testonly = True,  # Ensure alwayslink does not leak in the codebase.
     srcs = [
         "transforms/init_text_file_to_import_test_pass.cc",
         "transforms/initialize_variables_in_session_init_test_pass.cc",
@@ -1938,7 +1938,7 @@ cc_library(
     name = "error_util",
     srcs = ["utils/error_util.cc"],
     hdrs = ["utils/error_util.h"],
-    visibility = ["//tensorflow:__subpackages__"],
+    visibility = ["//visibility:public"],
     deps = [
         "//tensorflow/compiler/xla/mlir/utils:error_util",
         "//tensorflow/core/platform:errors",
@@ -2240,7 +2240,7 @@ cc_library(
 
 cc_library(
     name = "tf_xla_mlir_translate",
-    testonly = True,  # Ensure alwayslink does not leak in the codebase.
+    # testonly = True,  # Ensure alwayslink does not leak in the codebase.
     srcs = ["utils/tf_xla_mlir_translate.cc"],
     deps = [
         ":compile_mlir_util",
diff --git a/tensorflow/compiler/mlir/tools/kernel_gen/ir/BUILD b/tensorflow/compiler/mlir/tools/kernel_gen/ir/BUILD
index f040ca2af3b..2695155f0b9 100644
--- a/tensorflow/compiler/mlir/tools/kernel_gen/ir/BUILD
+++ b/tensorflow/compiler/mlir/tools/kernel_gen/ir/BUILD
@@ -5,9 +5,7 @@ load("//tensorflow:tensorflow.default.bzl", "get_compatible_with_cloud")
 package(
     # copybara:uncomment default_applicable_licenses = ["//tensorflow:license"],
     default_visibility = [
-        "//tensorflow/compiler/mlir/tools/kernel_gen:friends",
-        # Allow visibility from the mlir language server.
-        "//learning/brain/mlir/mlir_lsp_server:__pkg__",
+        "//visibility:public",
     ],
     licenses = ["notice"],
 )
diff --git a/tensorflow/compiler/mlir/tosa/BUILD b/tensorflow/compiler/mlir/tosa/BUILD
index b942c64f611..32584cafaee 100644
--- a/tensorflow/compiler/mlir/tosa/BUILD
+++ b/tensorflow/compiler/mlir/tosa/BUILD
@@ -9,7 +9,7 @@ load("@llvm-project//mlir:tblgen.bzl", "gentbl_cc_library")
 # TODO: Tighten visibility once targets are at the right granularity.
 package(
     # copybara:uncomment default_applicable_licenses = ["//tensorflow:license"],
-    default_visibility = [":internal"],
+    default_visibility = ["//visibility:public"],
     licenses = ["notice"],
 )
 
@@ -134,7 +134,6 @@ cc_library(
         "transforms/passes.h",
     ],
     compatible_with = get_compatible_with_cloud(),
-    visibility = [":friends"],
     deps = [
         ":legalize_common",
         ":passes_header",
@@ -187,7 +186,6 @@ cc_library(
         "transforms/passes.h",
     ],
     compatible_with = get_compatible_with_cloud(),
-    visibility = [":friends"],
     deps = [
         ":legalize_common",
         ":passes_header",
@@ -219,7 +217,6 @@ cc_library(
         "transforms/passes.h",
     ],
     compatible_with = get_compatible_with_cloud(),
-    visibility = [":friends"],
     deps = [
         ":legalize_common",
         ":passes_header",
diff --git a/tensorflow/compiler/tf2xla/BUILD b/tensorflow/compiler/tf2xla/BUILD
index bca5391f9f3..7e3b67864b2 100644
--- a/tensorflow/compiler/tf2xla/BUILD
+++ b/tensorflow/compiler/tf2xla/BUILD
@@ -21,7 +21,7 @@ load("//tensorflow/compiler/xla/service/cpu:build_defs.bzl", "runtime_copts")
 
 package(
     # copybara:uncomment default_applicable_licenses = ["//tensorflow:license"],
-    default_visibility = [":internal"],
+    default_visibility = ["//visibility:public"],
     licenses = ["notice"],
 )
 
@@ -635,7 +635,6 @@ cc_library(
     hdrs = [
         "xla_helpers.h",
     ],
-    visibility = [":friends"],
     deps = [
         ":common",
         ":host_compute_metadata_proto_cc",
@@ -717,7 +716,6 @@ cc_library(
         "shape_util.h",
         "type_util.h",
     ],
-    visibility = [":friends"],
     deps = [
         "//tensorflow/compiler/xla:literal",
         "//tensorflow/compiler/xla:shape_util",
diff --git a/tensorflow/compiler/xla/service/BUILD b/tensorflow/compiler/xla/service/BUILD
index 57555636cdc..2ec8e2ab672 100644
--- a/tensorflow/compiler/xla/service/BUILD
+++ b/tensorflow/compiler/xla/service/BUILD
@@ -30,7 +30,7 @@ load("//tensorflow/compiler/xla/service:xla_compile.bzl", "xla_aot_compile_cpu",
 
 package(
     # copybara:uncomment default_applicable_licenses = ["//tensorflow:license"],
-    default_visibility = [":friends"],
+    default_visibility = ["//visibility:public"],
     licenses = ["notice"],
 )
 
diff --git a/tensorflow/core/BUILD b/tensorflow/core/BUILD
index a8af5a0c6f8..8361ca0eafd 100644
--- a/tensorflow/core/BUILD
+++ b/tensorflow/core/BUILD
@@ -114,11 +114,7 @@ load(
 
 package(
     # copybara:uncomment default_applicable_licenses = ["//tensorflow:license"],
-    default_visibility = [
-        ":dependency_allowlist",
-        "//tensorflow:internal",
-        "//tensorflow_models:__subpackages__",
-    ],
+    default_visibility = ["//visibility:public"],
     features = if_google([
         "-layering_check",
         "-parse_headers",
diff --git a/tensorflow/core/common_runtime/BUILD b/tensorflow/core/common_runtime/BUILD
index eaed483119e..49dff78f0b0 100644
--- a/tensorflow/core/common_runtime/BUILD
+++ b/tensorflow/core/common_runtime/BUILD
@@ -43,7 +43,7 @@ default_package_visibility = [
 
 package(
     # copybara:uncomment default_applicable_licenses = ["//tensorflow:license"],
-    default_visibility = default_package_visibility,
+    default_visibility = ["//visibility:public"],
     features = if_google(
         [
             "-layering_check",
diff --git a/tensorflow/core/grappler/optimizers/data/BUILD b/tensorflow/core/grappler/optimizers/data/BUILD
index 7f8e775d34c..4882ae37463 100644
--- a/tensorflow/core/grappler/optimizers/data/BUILD
+++ b/tensorflow/core/grappler/optimizers/data/BUILD
@@ -5,11 +5,7 @@ load("//tensorflow/core/platform:build_config.bzl", "tf_protos_all")
 package(
     # copybara:uncomment default_applicable_licenses = ["//tensorflow:license"],
     default_visibility = [
-        "//tensorflow/core/data:__pkg__",
-        "//tensorflow/core/data/service:__pkg__",
-        "//tensorflow/core/grappler/optimizers/data:__subpackages__",
-        "//tensorflow/core/kernels/data:__pkg__",
-        "//tensorflow/core/kernels/data/experimental:__pkg__",
+        "//visibility:public"
     ],
     licenses = ["notice"],
 )
