diff --git a/tensorflow/compiler/mlir/lite/transforms/dilated_conv.h b/tensorflow/compiler/mlir/lite/transforms/dilated_conv.h
index fe8bb7d2ca1..38caa663968 100644
--- a/tensorflow/compiler/mlir/lite/transforms/dilated_conv.h
+++ b/tensorflow/compiler/mlir/lite/transforms/dilated_conv.h
@@ -113,9 +113,9 @@ LogicalResult ConvertTFDilatedConvOp<Conv2dOpTy>::matchAndRewrite(
   // Allow dynamic width and height dimensions only.
   auto result_ty = mlir::cast<TensorType>(op.getResult().getType());
   if (!result_ty.hasRank() || result_ty.getRank() != 4 ||
-      result_ty.isDynamicDim(0) || result_ty.isDynamicDim(3)) {
+      result_ty.isDynamicDim(3)) {
     return rewriter.notifyMatchFailure(
-        op, "only dynamic width and height dimensions are allowed");
+        op, "only dynamic batch, width and height dimensions are allowed");
   }
 
   // Check if the ConvOp's input is defined by `Expand` op, and the output used
