diff --git a/src/Dialect/Mlir/DialectBuilder.cpp b/src/Dialect/Mlir/DialectBuilder.cpp
index 593ab7a4..2f2a8f8a 100644
--- a/src/Dialect/Mlir/DialectBuilder.cpp
+++ b/src/Dialect/Mlir/DialectBuilder.cpp
@@ -479,6 +479,9 @@ Value MathBuilder::constantIndex(int64_t val) const {
 TypedAttr MathBuilder::negativeInfAttr(mlir::Type type) const {
   TypedAttr attr;
   TypeSwitch<Type>(type)
+      .Case<Float16Type>([&](Type) {
+        attr = b().getF16FloatAttr(-std::numeric_limits<float>::infinity());
+      })
       .Case<Float32Type>([&](Type) {
         attr = b().getF32FloatAttr(-std::numeric_limits<float>::infinity());
       })
@@ -524,6 +527,9 @@ TypedAttr MathBuilder::negativeInfAttr(mlir::Type type) const {
 TypedAttr MathBuilder::positiveInfAttr(mlir::Type type) const {
   TypedAttr attr;
   TypeSwitch<Type>(type)
+      .Case<Float16Type>([&](Type) {
+        attr = b().getF16FloatAttr(std::numeric_limits<float>::infinity());
+      })
       .Case<Float32Type>([&](Type) {
         attr = b().getF32FloatAttr(std::numeric_limits<float>::infinity());
       })
