diff --git a/src/Tools/onnx-mlir-opt/CMakeLists.txt b/src/Tools/onnx-mlir-opt/CMakeLists.txt
index a90a670a..0a80c88b 100644
--- a/src/Tools/onnx-mlir-opt/CMakeLists.txt
+++ b/src/Tools/onnx-mlir-opt/CMakeLists.txt
@@ -20,3 +20,17 @@ add_onnx_mlir_executable(onnx-mlir-opt
   MLIROptLib
   MLIRSCFToOpenMP
   )
+
+add_onnx_mlir_library(OMRegisterPasses
+  RegisterPasses.cpp
+
+  EXCLUDE_FROM_OM_LIBS
+
+  LINK_LIBS PUBLIC
+  OMCompilerPasses
+  OMAccelerator
+  MLIRAffineTransforms
+  MLIRLinalgTransforms
+  MLIRMemRefTransforms
+)
+
