diff --git a/src/Transform/ONNX/ONNXOpTransformPass.cpp b/src/Transform/ONNX/ONNXOpTransformPass.cpp
index aa627e73..8d36705e 100644
--- a/src/Transform/ONNX/ONNXOpTransformPass.cpp
+++ b/src/Transform/ONNX/ONNXOpTransformPass.cpp
@@ -79,13 +79,6 @@ void ONNXOpTransformPass::runOnOperation() {
   OperationFingerPrint before(module);
   do {
     OpPassManager dynamicPM("builtin.module");
-    dynamicPM.addNestedPass<func::FuncOp>(
-        onnx_mlir::createDecomposeONNXToONNXPass());
-    if (enableRecomposeOptPass)
-      dynamicPM.addNestedPass<func::FuncOp>(
-          onnx_mlir::createRecomposeONNXToONNXPass());
-    dynamicPM.addNestedPass<func::FuncOp>(
-        onnx_mlir::createShapeInferencePass());
     dynamicPM.addPass(mlir::createCanonicalizerPass());
     dynamicPM.addNestedPass<func::FuncOp>(
         onnx_mlir::createShapeInferencePass());
